man/notify-selfcontained-example.c

```diff
@@ -0,0 +1,184 @@
+/* SPDX-License-Identifier: MIT-0 */
+
+/* Implement the systemd notify protocol without external dependencies.
+ * Supports both readiness notification on startup and on reloading,
+ * according to the protocol defined at:
+ * https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html
+ * This protocol is guaranteed to be stable as per:
+ * https://systemd.io/PORTABILITY_AND_STABILITY/ */
+
+#include <assert.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <signal.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <time.h>
+#include <unistd.h>
+
+#define _cleanup_(f) __attribute__((cleanup(f)))
+
+static volatile sig_atomic_t reloading = 0;
+static volatile sig_atomic_t terminating = 0;
+
+static void closep(int *fd) {
+  if (!fd || *fd < 0)
+    return;
+
+  close(*fd);
+  *fd = -1;
+}
+
+static int notify(const char *message) {
+  union sockaddr_union {
+    struct sockaddr sa;
+    struct sockaddr_un sun;
+  } socket_addr = {
+    .sun.sun_family = AF_UNIX,
+  };
+  size_t path_length, message_length;
+  _cleanup_(closep) int fd = -1;
+  const char *socket_path;
+
+  assert(message);
+
+  message_length = strlen(message);
+  if (message_length == 0) {
+    fprintf(stderr, "Message to send to $NOTIFY_SOCKET is empty\n");
+    return -EINVAL;
+  }
+
+  socket_path = getenv("NOTIFY_SOCKET");
+  if (!socket_path)
+    return 0; /* Not running under systemd? Nothing to do */
+
+  /* Only AF_UNIX is supported, with path or abstract sockets */
+  if (socket_path[0] != '/' && socket_path[0] != '@')
+    return -EAFNOSUPPORT;
+
+  path_length = strlen(socket_path);
+  /* Ensure there is room for NUL byte */
+  if (path_length >= sizeof( socket_addr.sun.sun_path)) {
+    fprintf(stderr, "$NOTIFY_SOCKET contains a path that is too long for AF_UNIX\n");
+    return -E2BIG;
+  }
+
+  memcpy(socket_addr.sun.sun_path, socket_path, path_length);
+
+  /* Support for abstrace socket */
+  if (socket_addr.sun.sun_path[0] == '@')
+    socket_addr.sun.sun_path[0] = 0;
+
+  fd = socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0);
+  if (fd < 0) {
+    fprintf(stderr, "Failed to open $NOTIFY_SOCKET\n");
+    return -errno;
+  }
+
+  if (connect(fd, &socket_addr.sa, sizeof(socket_addr)) != 0) {
+    fprintf(stderr, "Failed to connect to $NOTIFY_SOCKET\n");
+    return -errno;
+  }
+
+  ssize_t written = write(fd, message, message_length);
+  if (written != (ssize_t) message_length) {
+    fprintf(stderr, "Failed to write message to $NOTIFY_SOCKET\n");
+    return written < 0 ? -errno : -EPROTO;
+  }
+
+  return 1; /* Notified! */
+}
+
+static int notify_ready() {
+  return notify("READY=1");
+}
+
+static int notify_reloading() {
+  char reload_message[sizeof("RELOADING=1\nMONOTONIC_USEC=") + 20 + 1];
```

Comment:
Why `+ 1`?

Maybe write this as
```C
/* A buffer with length sufficient to format the maximum UINT64 value */
char reload_message[sizeof("RELOADING=1\nMONOTONIC_USEC=18446744073709551615")];
```

Reply:
For the trailing null

Reply:
That's already included in retval of `sizeof()`?